File: C/linked_list.c
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>

/*	
	Linked List is a linear data structure. 
	It is a collection of nodes that contain data and link for next node.
	
	The common structure of a linked list node is:
	
	[int data, LinkedList next] : * where data stores the value of the node 
				   	   		 	  * next contains the address of the value after the current
				       		 	    node in the list.

*/

//Definition of a node in Linked List 

struct node
{
	int value;  
	struct Node *next;
};


typedef struct node List;  /*Instead of writing the complete definition of the node everytime
							   we can provide it a new name for easier implementation*/

List* head = NULL;	// Variable to store the head of Linked List. DO NOT CHANGE UNTIL THERE IS CHANGE IN THE HEAD NODES.

List* insertAtBeginning(List*, int); //Function to insert node at the beginning of list.
List* insertAtLast(List*, int); //Function to insert node at the of list.
List* display(List*); 			//Function to display the values of list in sequential order. 
List* deleteFirstNode(List*);	//Function to delete first node from the list.
List* deleteByValue(List*, int);	  	//Function to delete a node by value from the list.
List* deleteLastNode(List*); 	//Function to delete last node from the list.

int main() 				  //Driver Code
{
	int value, option;  //Option stores the id of operation user wants to perform
	do			// Operation menu
	{
		
		printf("\n\n\t*****MAIN MENU*****\n");
		printf("\n\t1. INSERT AT BEGINNING");
		printf("\n\t2. INSERT AT LAST");
		printf("\n\t3. DELETE FIRST NODE");
		printf("\n\t4. DELETE LAST NODE");
		printf("\n\t5. DELETE BY VALUE");
		printf("\n\t6. DISPLAY LIST");
		printf("\n\t7. EXIT");
		printf("\n\tEnter your option : ");
		scanf("%d", &option);
		switch(option)
		{
		
			case 1:
				printf("\n\tEnter the value to be added in the list : ");
				scanf("%d", &value); //value contains the value to be inserted at the beginning
				head = insertAtBeginning(head, value); 
				break;
				
			case 2:
				printf("\n\tEnter the value to be added in the list : ");
				scanf("%d", &value); //value contains the value to be inserted at the end
				head = insertAtLast(head, value); 
				break;
				
			case 3:
				head = deleteFirstNode(head);
				break;
				
			case 4:
				head = deleteLastNode(head);
				break;
				
			case 5:
				printf("\n\tEnter the value to be deleted from the list : ");
				scanf("%d", &value); //value contains the value to be deleted
				head = deleteByValue(head, value);
				break;
				
			case 6: 
				head = display(head);
				break;
			
			case 7: break;
			
			default: printf("\n\tINVALID CHOICE INPUT!!!\n");
		}
	}while(option != 7);
	return 0;
}

List* insertAtBeginning(List* listHead, int value)
{
	/*
		Create a new node. 
		Assign value to the node.
		Assign head->next = new node 
	*/
	
	List* newNode;
	newNode = (List*)malloc(sizeof(List));
	
	newNode -> value = value;
	newNode -> next = listHead;
	listHead = newNode;
	
	return listHead;
}

List* insertAtLast(List* listHead, int value)
{
	/*
		Create a new node. 
		Assign value to the node.
		Check if the list is empty.
		If true, make the new node the first node and assign head pointer to it
		Else, traverse to the last node and set ptr->next - new node
	*/
	List* ptr = listHead;
	List* newNode;
	newNode = (List*)malloc(sizeof(List));
	
	newNode -> value = value;
	newNode -> next = NULL;
	
	if(ptr==NULL)
		listHead = newNode;

	else
	{
		while(ptr->next != NULL)
			ptr = ptr->next;
		
		ptr->next = newNode;
	}
	return listHead;
}

List* deleteFirstNode(List* listHead)
{
	/*
		First check for empty list.
		If not empty, set a new pointer ptr to the head node.
		Shift the head to next node.
		Free ptr.
		
		Shifting the head pointer to next node handles the problem of deleted node being the first node
		as it allocates NULL to the head node.
	*/
	List* ptr = listHead;
	if(ptr == NULL)
		printf("\n\tEMPTY LIST!\n");
	else
	{
		listHead = listHead -> next;
		printf("\n\tThe value being deleted is: %d\n", ptr -> value);
		free(ptr);
	}
	return listHead;
}

List* deleteLastNode(List* listHead)
{
	/*
		First check for empty list.
		If not empty, set a new pointer ptr to the head node.
		Traverse to the last node. 
		If head node is the last node, set head node = NULL
		Free ptr

	*/
	List* ptr = listHead;
	if(ptr == NULL)
		printf("\n\tEMPTY LIST!\n");
	else
	{
		while(ptr->next != NULL)
			ptr = ptr->next;
			
		printf("\n\tThe value being deleted is: %d\n", ptr -> value);
		if(ptr==listHead)
			listHead = NULL;
		free(ptr);
	}
	return listHead;
	
}

List* deleteByValue(List* listHead, int value)
{
	/*
		First check for empty list.
		If not empty, set new pointers ptr and temp to the head node.
		
		Check if value to be deleted is at the head node, 
		If true, follow the steps of deleting first node
		Else, traverse temp to the node containing the value and ptr to the prev node
			Check if temp is NULL, if true, display ELEMENT ABSENT
			Else, set ptr->next = temp->next and free temp

	*/
	List* ptr = listHead;
	List* temp = listHead; 
	
	if(ptr==NULL)
		printf("\n\tEMPTY LIST!\n");
	
	else if(ptr != NULL && listHead->value == value)
	{
		printf("\n\tThe value being deleted is: %d\n", ptr->value);
		listHead = listHead->next;
		free(ptr);
	}

	else
	{
		while (temp != NULL && temp->value != value) 
		{
	        ptr = temp;
	        temp = temp->next;
    	}
    	if(temp==NULL)
    		printf("\n\tVALUE %d ABSENT FROM LIST!\n", value);
    	else
    	{
			printf("\n\tThe value being deleted is: %d\n", temp->value);
			ptr->next = temp->next;
			free(temp);		
		}
		
	}
	return listHead;
}

List* display(List* listHead)
{
		/* 
		To display, first check for empty list, 
		If not empty, traverse sequentially until ptr = NULL and display the values sequentially.
	*/
	
	List *ptr = listHead;
	if(ptr == NULL) 
		printf("\n\tEMPTY LIST!\n");
	
	else
	{
		printf("\n");
		while(ptr != NULL)
		{
			printf("\t%d", ptr -> value);
			ptr = ptr -> next;
		}
	}
	return listHead;
}

File: Java/Data Structures/LinkedList.java
import java.util.*;
public class LinkedList 
{
    ListNode head, tail;
    LinkedList()
    {
        head = tail = null;
    }
    private void createList(int[] A)
    {
        for(int val : A)
            insertLast(val);
    }
    private void insertBeginning(int val)
    {
        ListNode temp = new ListNode(val);
        
        if(head == null)
        {
            head = temp; 
            tail = head;
        }
        else
        {
            temp.next = head; 
            head = temp;
        }
    }
    private void insertLast(int val)
    {
        ListNode temp = new ListNode(val);
        
        if(head == null)
        {
            head = temp; 
            tail = head;
        }
        else
        {
            tail.next = temp;
            tail = tail.next;
        }
    }
    private int deleteFirst()
    {
        if(head == null)
            return -1; 

        int value = head.data;

        if(head == tail)
        {
            head = tail = null; 
            return value;
        }

        ListNode curr = head; 
        head = head.next; 
        curr.next = null;
        return value;
    }
    private int deleteLast()
    {
        if(head == null)
            return -1; 

        int value = head.data;

        if(head == tail)
        {
            head = tail = null; 
            return value;
        }

        ListNode curr = head; 
        if(head == tail)
        
        while(curr.next.next != null)
            curr = curr.next;

        value = curr.next.data; 
        tail = curr;
        curr.next = null;
        return value;
    }
    private void display()
    {
        if(head == null)
            System.out.println("EMPTY LIST");

        else
        {
            ListNode curr = head; 
            while(curr != null)
            {
                System.out.print(curr.data+" ");
                curr = curr.next; 
            }
            System.out.println();
        }
    }
    public static void main(String[] args) 
    {
        Scanner ob = new Scanner(System.in);
        LinkedList ll = new LinkedList();

        System.out.println("1. Insert Beginning");
        System.out.println("2. Insert Last");
        System.out.println("3. Create list from array");
        System.out.println("4. Delete Beginning");
        System.out.println("5. Delete Last");
        System.out.println("6. Display");
        System.out.println("7. Exit");
    
        while(true)
        {
            System.out.print("Enter your choice : ");
            int ch = ob.nextInt();

            switch(ch)
            {
                case 1:
                    System.out.print("Enter data : ");
                    int data1 = ob.nextInt();
                    ll.insertBeginning(data1);
                    break;

                case 2:
                    System.out.print("Enter data : ");
                    int data2 = ob.nextInt();
                    ll.insertLast(data2);
                    break;
                
                case 3:
                    System.out.print("Enter array size : ");
                    int n = ob.nextInt(); 
                    int A[] = new int[n];

                    System.out.print("Enter array elements : ");
                    for(int i=0; i<n; i++)
                        A[i] = ob.nextInt();
                    
                    ll.createList(A);
                    break; 

                case 4:
                    int val1 = ll.deleteFirst();
                    if(val1 == -1)
                        System.out.println("EMPTY LIST");
                    else
                        System.out.println("DELETED VALUE : " + val1);   
                    break; 

                case 5:
                    int val2 = ll.deleteLast();
                    if(val2 == -1)
                        System.out.println("EMPTY LIST");
                    else
                        System.out.println("DELETED VALUE : " + val2);   
                    break; 

                case 6:
                    ll.display();
                    break;
                
                case 7:
                    ob.close();
                    System.exit(0);

                default:
                    System.out.println("INVALID INPUT");            
            }
        }   
    }
}

